[WIP] Add inbound and outbound relaying via transport maps

SamuelNitsche opened the pull request:
I have also worked on a solution to implement inbound and outbound relaying via transport maps. I've implemented all the functions so far, but I haven't completely tested them yet. I have seen that [nupplaphil](https://github.com/nupplaphil) has also been working on a solution, but I think his implementation is much too complex for a mail server with a lot of email traffic. Please note that this pull request is a work in progress solution.

SamuelNitsche committed:
Add transport routes database

SamuelNitsche committed:
Added postfix routing possibility

SamuelNitsche committed:
Changes for relaying

SamuelNitsche committed:
Started implementing relay domains UI

SamuelNitsche committed:
UI

SamuelNitsche committed:
Load table

SamuelNitsche committed:
UI

SamuelNitsche committed:
UI

SamuelNitsche committed:
UI

SamuelNitsche committed:
Debugging

SamuelNitsche committed:
UI

SamuelNitsche committed:
UI

SamuelNitsche committed:
Added mass action option

SamuelNitsche committed:
UI

SamuelNitsche committed:
Added UI edit page

SamuelNitsche committed:
Fixed php syntax

SamuelNitsche committed:
Added values

SamuelNitsche committed:
Added sql code to insert relay domain

SamuelNitsche committed:
.

SamuelNitsche committed:
Deleting maps

SamuelNitsche committed:
Editing maps

SamuelNitsche committed:
fixes

SamuelNitsche committed:
Fixes

SamuelNitsche committed:
Disabled redirects

SamuelNitsche committed:
debugging

SamuelNitsche committed:
Added missing methods

SamuelNitsche committed:
uncommented redirects

SamuelNitsche committed:
Fixed debugging

SamuelNitsche committed:
Fixed mysql syntax

SamuelNitsche committed:
Fixed mysql query

SamuelNitsche committed:
Fixed edit transpot route

SamuelNitsche committed:
Added semicolon

SamuelNitsche committed:
Added localization

SamuelNitsche committed:
Replaced by lorem ipsum

Commit:
Merge branch 'master' of https://github.com/SamuelNitsche/mailcow-dockerized

Merge

Commit:
English locale for relay

Commit:
Added created, modified, active columns

Commit:
Reflected database changes

Commit:
Reflected database changes

Commit:
Added active toggles to frontend

Commit:
Fixed status when editing

Commit:
Added localized labels

Commit:
Added domain select

Commit:
Fixed modal

Commit:
Fixed select name

Commit:
Bugfix

Commit:
Modal bugfix

Commit:
Modal bugfix

Commit:
Fixed select all button in relay table

Commit:
Added enable/disable functionality to transport_map table

Commit:
Added some validation

Commit:
Refactored methods an naming

Commit:
Changed db column name

Commit:
Fixed tab link

Commit:
Naming conventions

Commit:
Add default value

Commit:
Changed name of function file according to function name

Commit:
Fixed form id

Commit:
Adjusted names and fixed button

Commit:
Dockerfile

Commit:
Revert #784b373

Commit:
Merge branch 'master' of git://github.com/mailcow/mailcow-dockerized

droid-sheep committed:
removed encoding for data attribute to fix deletion

andryyy committed:
Merge pull request #1148 from droid-sheep/master

[Web] remove encoding for resource data attribute to fix deletion

andryyy committed:
[Web] Fix empty exlude regex in sync job

andryyy committed:
Merge branch 'master' of https://github.com/mailcow/mailcow-dockerized

SamuelNitsche committed:
Merged master branch

SamuelNitsche committed:
Fixed merge conflict

SamuelNitsche committed:
Removed reference of not existing file

nupplaphil commented:
[Edit] Removed first post
As I see that @andryyy is working on this PR too, I think it's better to work on this instead of working on my PR. I will close my PR ;-)

SamuelNitsche commented:
As I already wrote, this pull request is not fully tested. Also there are few spots where I have to replace my placeholder texts. I absolutely didn‘t want to start a competition. I just wanted to show that there is another solution. :)

nupplaphil commented:
@SamuelNitsche Great idea to transport mails of whole domains and specific mailboxes too! I switched to your master and it seems to work for me now.

--> Next week I will try to use mailbox-specific transport_maps

[Edit] One suggestion: I have to make the nexthop like "smtp:IP:port". I think it's better to either describe how to use the nexthop field or to make different textboxes (maybe with a dropdown for smtp/smtps)

SamuelNitsche commented:
Good catch! I think adding different input fields is a good idea. Thanks!

[Edit] I have now implemented different input boxes, a basic description and some help texts. Also I have tested the whole system a bit more and for me it is working perfectly fine.

SamuelNitsche committed:
Removed bcc tab

SamuelNitsche committed:
Split nexthop field into different inputs

SamuelNitsche committed:
Fixed styling of input fields

SamuelNitsche committed:
Backend implementation of splitted fields

SamuelNitsche committed:
Added colon between ip and port

SamuelNitsche committed:
Split nexthop field into different inputs

SamuelNitsche committed:
Don't add colon when port is empty

SamuelNitsche committed:
Display correct values

SamuelNitsche committed:
Backend implementation of splitted fields

SamuelNitsche committed:
Adjusted info texts

SamuelNitsche committed:
Typo

SamuelNitsche committed:
Added lmtps protocol

SamuelNitsche committed:
Added lmtps protocol

SamuelNitsche committed:
Add helper text to edit page

SamuelNitsche committed:
Added some browser input validation

SamuelNitsche committed:
WIP

andryyy commented:
Hm, but shouldn't it be using relayhosts we added via UI?

nupplaphil commented:
Hm, you said at #1068 that forwarding isn't the same as relayhosts. I think we shouldn't cross-use these functions.

nupplaphil commented:
I'd love to use the dev/master branch again. @andryyy what do Samuel/we have to do to geht merged? :-)

SamuelNitsche committed:
New relay implementation (#1)

* Merged mailbox-routes

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

SamuelNitsche commented:
We had another long and intensive internal discussion about the implementation of transport routes and now think that this can only be solved reasonably if we also solve issue #1068. Therefore I have updated the pull request and changed the way incoming mail routing works. The system has just passed all the internal tests. :-)
@andryyy The next two days I don't have not enough time to explain exactly what this update does. If you have any questions why and how we changed things, feel free to ask.

nupplaphil commented:
I'm sorry to repeat myself :-( , but are there any news so far?

SamuelNitsche committed:
Removed commented lines

SamuelNitsche commented on data/Dockerfiles/postfix/postfix.sh:
```diff
@@ -8,24 +8,6 @@ if [[ -z $(grep null /etc/aliases) ]]; then
-  )
-  END AS result;
-EOF
-
```
This query is no longer necessary, since it has been replaced by others.
(see main.cf -> mysql_relay_recipient_maps)

SamuelNitsche commented on data/Dockerfiles/postfix/postfix.sh:
```diff
@@ -43,6 +25,14 @@ query = SELECT IF(EXISTS(
+dbname = ${DBNAME}
+query = SELECT nexthop FROM transport_maps WHERE local_dest = '%s' AND active = '1';
+EOF
+
```
Returns the transport routes for mailboxes.

SamuelNitsche commented on data/Dockerfiles/postfix/postfix.sh:
```diff
@@ -160,28 +150,50 @@ user = ${DBUSER}
+        WHERE alias_domain='%s'
+          AND active='1';
 EOF
 
```
This query has been simplified since the relay host implementation has changed.

SamuelNitsche commented on data/Dockerfiles/postfix/postfix.sh:
```diff
@@ -160,28 +150,50 @@ user = ${DBUSER}
+query = SELECT nexthop FROM domain
+        WHERE domain='%s'
+          AND active = '1';
+EOF
```
Returns the transport routes for whole domains.

SamuelNitsche commented on data/Dockerfiles/postfix/postfix.sh:
```diff
@@ -160,28 +150,50 @@ user = ${DBUSER}
+        WHERE domain='%d'
+          AND nexthop <> 'lmtp:[dovecot]'
+          AND active = '1';
 EOF
```
Returns all domains that should be forwarded to other mail servers via transport maps. Note that the 'relay_all_recipients' column has been replaced by defining a non default nexthop.

SamuelNitsche commented on data/conf/postfix/main.cf:
```diff
@@ -63,7 +64,7 @@ proxy_read_maps = proxy:mysql:/opt/postfix/conf/sql/mysql_virtual_sender_acl.cf,
 queue_run_delay = 300s
 relay_domains = proxy:mysql:/opt/postfix/conf/sql/mysql_virtual_relay_domain_maps.cf
-relay_recipient_maps = proxy:mysql:/opt/postfix/conf/sql/mysql_relay_recipient_maps.cf
+relay_recipient_maps = $virtual_mailbox_maps $virtual_alias_maps proxy:mysql:/opt/postfix/conf/sql/mysql_transport_all_recipients_maps.cf proxy:mysql:/opt/postfix/conf/sql/mysql_transport_maps.cf
```
See the comment above.

SamuelNitsche commented on data/conf/postfix/main.cf:
```diff
@@ -106,12 +107,13 @@ virtual_alias_maps = proxy:mysql:/opt/postfix/conf/sql/mysql_virtual_alias_maps.
 virtual_gid_maps = static:5000
 virtual_mailbox_base = /var/vmail/
-virtual_mailbox_domains = proxy:mysql:/opt/postfix/conf/sql/mysql_virtual_domains_maps.cf
+virtual_alias_domains = proxy:mysql:/opt/postfix/conf/sql/mysql_virtual_domains_maps.cf
```
This implementation is technically not wrong, but better solved in this way.
[Postfix virtual_mailbox_domains](http://www.postfix.org/postconf.5.html#virtual_mailbox_domains)
[Postfix virtual_alias_domains](http://www.postfix.org/postconf.5.html#virtual_alias_domains)

SamuelNitsche commented:
Sorry for the little spam. I added some comments to my changes to (hopefully) explain more detailed why and what changes were made.
@andryyy I'd appreciate it if you'd look at the changes and leave your opinion or wishes. Thanks!

nupplaphil commented:
@andryyy @SamuelNitsche

I'm sorry to ask again, but do you think you can merge it? :-( I don't do any upgrades since a month for mailcow because of this missing feature

andryyy commented:
This is something with very special changes that will break a lot of installations . I am also very sure some stuff is not 100% tested and will fail. Some configs values/parameters were chosen for a good reason and are now replaced to fix the special scenario this PR was made for.

I just don't have the time to check if everything still works as intended.

andryyy commented:
It will definitely destroy working backup mx setups, it does not migrate settings. I'm not sure about relayed users. It seems to just use transport_maps for everything now and kind of misuses relay rcpt maps. What about relayhosts? Do they still work? I didn't check all of it, this is still a very good change - don't get me wrong - but there is much more to it. :-(

nupplaphil commented:
Oh..

Sounds like a LOT of work. So was #1136 a more comfortable way for it with less side effects? I "just" reused the relayhost for transport_maps and didn't change anything else.

Or what about @SamuelNitsche first change before he completely changed the strategy for the transport_maps ? 

I don't want to stress, but for me it seems that this is a long run PR with a good chance that other changes will break this changes a lot more and make it more and more unlikely to get merged.

Or what do you think?

SamuelNitsche commented:
Since our ideas obviously differ, it probably makes more sense to close this pull request. It is probably best if you integrate this function as it fits into your existing concept.
